atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CellReference;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.common.streams.KeyedStream;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+public final class TransactionScopedCacheImpl implements TransactionScopedCache {
+    private final TransactionCacheValueStore valueStore;
+
+    public TransactionScopedCacheImpl(ValueCacheSnapshot snapshot) {
+        valueStore = new TransactionCacheValueStoreImpl(snapshot);
+    }
+
+    // TODO(jshah): figure out how we can improve perf given that this is now synchronised. Maybe its fine, but maybe
+    //  we should instead autobatch?
+    @Override
+    public synchronized void write(TableReference tableReference, Cell cell, CacheValue value) {
+        valueStore.cacheLocalWrite(tableReference, cell, value);
+    }
+
+    // TODO(jshah): as above. Equally, we should probably use the async value loader, then have it get afterwards,
+    //  thus reducing the time spent in this method.
```

Comment:
I quite frankly couldn't motivate myself to wire the async stuff in this PR as I think that'd just make the complexity insane when taken in all at once.